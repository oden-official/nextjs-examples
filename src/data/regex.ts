export const REG_EXP = {
  EMAIL: /(([^<>()\[\]\\.,;:\s@"]+(\.[^<>()\[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))/,
  PW: /^(?=.*?[a-zA-Z])(?=.*?[0-9])(?=.*?[#?!@$ %^&*-]).{8,20}$/
}