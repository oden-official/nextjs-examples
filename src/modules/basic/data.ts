export const EMAIL_ERR = {
  NOT_VALID: "이메일 양식이 맞지 않습니다."
}

export const PW_ERR = {
  NOT_VALID: "비밀번호 양식이 맞지 않습니다.",
  NOT_MATCH: "비밀번호가 일치하지 않습니다."
}