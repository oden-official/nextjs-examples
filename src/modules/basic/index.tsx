import InputTextTopLabeled from '../../components/atom/input-text-top-labeled';
import styles from './styles.module.scss';
import { useLayer } from './use-layer';

const BasicModule: React.FunctionComponent = () => {
  const {
    email,
    emailErr,
    pw,
    pwConfirm,
    pwErr,
    ok,

    onChangeEmail,
    onChangePw,
    onChangePwConfirm,
  } = useLayer();

  return <div className={styles.main}>
    <div role={'test'}>test</div>
    <InputTextTopLabeled
      label={'이메일'}
      id={'email'}
      className={styles.input}
      inputAttr={{
        value: email,
        onChange: onChangeEmail,
        role: 'email',
      }}
    />
    <div role='emailErr'>{emailErr}</div>
    <InputTextTopLabeled
      label={'비밀번호'}
      id={'pw'}
      className={styles.input}
      inputAttr={{
        value: pw,
        onChange: onChangePw,
        role: 'pw',
      }}
    />
    <InputTextTopLabeled
      label={'비밀번호 확인'}
      id={'pwConfirm'}
      className={styles.input}
      inputAttr={{
        value: pwConfirm,
        onChange: onChangePwConfirm,
        role: 'pwConfirm',
      }}
    />
    <div>{pwErr}</div>
    <div>{ok ? 'ok' : 'disabled'}</div>
  </div>
}

export default BasicModule;
