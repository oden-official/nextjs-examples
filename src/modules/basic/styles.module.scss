.main {
  display: flex;
  align-items: center;
  flex-direction: column;
  margin-top: 3rem;
}

.input {
  margin-bottom: 1rem;
}