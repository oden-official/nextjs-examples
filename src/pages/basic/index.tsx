import { NextPage } from 'next';
import BasicModule from '../../modules/basic';

const Basic: NextPage = () => {
  return <BasicModule />
}

export default Basic;
