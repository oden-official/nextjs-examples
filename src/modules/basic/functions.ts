import { REG_EXP } from '../../data/regex';
import { EMAIL_ERR, PW_ERR } from './data';

export const emailErrHandler = (email: string) => {
  if (email === '') return '';

  const ok = REG_EXP.EMAIL.test(email);
  if (!ok) return EMAIL_ERR.NOT_VALID;
  return null;
}

export const pwErrHandler = (pw: string, pwConfirm: string) => {
  if (pw === '' && pwConfirm === '') return '';

  const ok = REG_EXP.PW.test(pw);
  if (!ok) return PW_ERR.NOT_VALID;

  if (pw !== pwConfirm) return PW_ERR.NOT_MATCH;
  
  return null;
}
