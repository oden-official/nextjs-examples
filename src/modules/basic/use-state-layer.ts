import { useInputTextWithErr } from "src/hooks/use-input-text-with-err";
import { emailErrHandler, pwErrHandler } from './functions';
import { useInputPwWithErr } from "src/hooks/use-input-pw-with-err";

export const useStateLayer = () => {
  const [email, onChangeEmail, emailErr] = useInputTextWithErr('', emailErrHandler);
  const [[pw, onChangePw], [pwConfirm, onChangePwConfirm], pwErr] = useInputPwWithErr(['', ''], pwErrHandler);

  const ok = emailErr === null && pwErr === null;
  return {
    email,
    emailErr,
    pw,
    pwConfirm,
    pwErr,
    ok,

    onChangeEmail,
    onChangePw,
    onChangePwConfirm,
  }
}