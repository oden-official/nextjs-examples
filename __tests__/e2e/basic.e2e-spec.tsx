import { fireEvent, render, screen } from '@testing-library/react'
import BasicModule from 'src/modules/basic'
import '@testing-library/jest-dom'
import { EMAIL_ERR } from 'src/modules/basic/data';

describe("BasicModule", () => {
  let container: HTMLElement;
  beforeEach(() => {
    container = render(<BasicModule />).container;
  })

  it("snapshot", () => {
    expect(container).toMatchSnapshot()
  })

  describe("email input", () => {
    it(`email ok ====>>>>> null`, async () => {
      const emailEl = screen.getByRole('email')
      fireEvent.change(emailEl, { target: { value: "test@test.co.kr" } });
      
      const emailErrEl = screen.getByRole('emailErr');
      expect(emailErrEl).toBeEmptyDOMElement();
    })
  
    it(`email Err ====>>>>> text should be ${EMAIL_ERR.NOT_VALID}`, async () => {
      const emailEl = screen.getByRole('email')
      fireEvent.change(emailEl, { target: { value: "test" } });
      
      const emailErrEl = screen.getByRole('emailErr');
      expect(emailErrEl).toHaveTextContent(EMAIL_ERR.NOT_VALID);
    })
  })
  
})