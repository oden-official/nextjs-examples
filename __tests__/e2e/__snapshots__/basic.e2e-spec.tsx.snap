// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BasicModule snapshot 1`] = `
<div>
  <div
    class="main"
  >
    <div
      role="test"
    >
      test
    </div>
    <div
      class="box input"
    >
      <label
        class="label"
        for="email"
      >
        이메일
      </label>
      <input
        class="input"
        id="email"
        role="email"
        type="text"
        value=""
      />
    </div>
    <div
      role="emailErr"
    />
    <div
      class="box input"
    >
      <label
        class="label"
        for="pw"
      >
        비밀번호
      </label>
      <input
        class="input"
        id="pw"
        role="pw"
        type="text"
        value=""
      />
    </div>
    <div
      class="box input"
    >
      <label
        class="label"
        for="pwConfirm"
      >
        비밀번호 확인
      </label>
      <input
        class="input"
        id="pwConfirm"
        role="pwConfirm"
        type="text"
        value=""
      />
    </div>
    <div />
    <div>
      disabled
    </div>
  </div>
</div>
`;
